processor/servicegraphprocessor/processor.go

```diff
@@ -288,7 +304,23 @@ func (p *serviceGraphProcessor) onExpire(e *store.Edge) {
 		zap.String("connection_type", string(e.ConnectionType)),
 		zap.Stringer("trace_id", e.TraceID),
 	)
+
 	stats.Record(context.Background(), statExpiredEdges.M(1))
+
+	if p.config.VirtualNodeFeatureEnabled {
+		// speculate virtual node before edge get expired.
+		if len(e.ClientService) == 0 {
+			e.ClientService = "user"
+		}
+
+		if len(e.ServerService) == 0 {
+			e.ServerService = p.getPeerHost(PeerAttributes, e.Peer)
+		}
+
+		e.ConnectionType = store.VirtualNode
+
+		p.store.OnComplete(e)
```

Comment:
The processor's `onComplete` and the store's `OnComplete` are essentially the same function. We can use the processor's here and unexport `OnComplete` and `OnExpire` in the store.

Reply:
That's a good point. Another alternative is simply calling `p.aggregateMetricsForEdge(e)`.